### PR TITLE
Allow user-defined user agents

### PR DESCRIPTION
**Pull request by jeskew**

### CODE REVIEW

File: src/ClientResolver.php [hunk 1]
```diff
@@ -146,7 +147,14 @@ class ClientResolver
             'doc'      => 'A handler that accepts a command object, request object and returns a promise that is fulfilled with an Aws\ResultInterface object or rejected with an Aws\Exception\AwsException. A handler does not accept a next handler as it is terminal and expected to fulfill a command. If no handler is provided, a default Guzzle handler will be utilized.',
             'fn'       => [__CLASS__, '_apply_handler'],
             'default'  => [__CLASS__, '_default_handler']
-        ]
+        ],
+        'user-agent' => [
```
Comment (jeremeamia) [inline src/ClientResolver.php]: We used underscores instead of hyphens for other config options.

File: src/ClientResolver.php [hunk 2]
```diff
@@ -441,6 +449,36 @@ public static function _apply_http_handler($value, array &$args, HandlerList $li
...
+        if (!is_array($value)) {
+            $value = [$value];
+        }
+
+        $value = array_map(
+            function ($ua) { return (string) $ua; },
+            $value
+        );
+
+        $value []= 'aws-sdk-php/' . Sdk::VERSION;
```
Comment (jeremeamia) [inline src/ClientResolver.php]: aws-sdk-php/{version} needs to be the first part.

**Comment by jeremeamia** [comment]
I'm not sure we should let people override the whole UA string. We always want `aws-sdk-php/VERSION` as the first part of the UA string.

File: src/ClientResolver.php [hunk 3]
```diff
@@ -441,6 +449,36 @@ public static function _apply_http_handler($value, array &$args, HandlerList $li
         );
     }
 
+    public static function _apply_user_agent($value, array &$args, HandlerList $list)
+    {
+        if (!is_array($value)) {
+            $value = [$value];
+        }
+
+        $value = array_map(
```
Comment (mtdowling) [inline src/ClientResolver.php]: I wonder if `$value = array_map('strval', $value);` would be faster/more clear?

**Comment by mtdowling** [comment]
This needs to be documented before merging, but otherwise LGTM

**Comment by jeremeamia** [comment]
:ship:

**Pull request merged**